src/ai/backend/manager/idle.py

```diff
@@ -326,6 +394,171 @@ async def update_app_streaming_status(
+        # Example: {cuda.device: 0, cuda.shares: 0.5} -> {cuda: 0.5}.
+        unique_res_map: DefaultDict[str, Any] = defaultdict(Decimal)
+        for k, v in occupied_slots.items():
+            unique_key = k.split('.')[0]
+            unique_res_map[unique_key] += v
+
+        # Do not take into account unallocated resources. For example, do not garbage collect
+        # a session without GPU even if cuda_util is configured in resource-thresholds.
+        for slot in unique_res_map:
+            if unique_res_map[slot] == 0:
+                unavailable_resources.update(self.slot_resource_map[slot])
+
+        # Respect idle_timeout, from keypair resource policy, over time_window.
+        policy_cache = self._policy_cache.get()
+        policy = policy_cache.get(session["access_key"], None)
+
+        if policy is None:
+            query = (
+                sa.select([keypair_resource_policies])
+                .select_from(
+                    sa.join(
+                        keypairs,
+                        keypair_resource_policies,
+                        (
+                            keypair_resource_policies.c.name
+                            == keypairs.c.resource_policy
+                        ),
+                    )
+                )
+                .where(keypairs.c.access_key == session["access_key"])
+            )
+            result = await dbconn.execute(query)
+            policy = result.first()
+            assert policy is not None
+            policy_cache[session["access_key"]] = policy
+        if policy["idle_timeout"] >= 0:
+            window_size = int(float(policy["idle_timeout"]) / interval)
+        if (window_size <= 0) or (math.isinf(window_size) and window_size > 0):
+            return True
+
+        # Get current utilization data from all containers of the session.
+        if session["cluster_size"] > 1:
+            query = (
+                sa.select([kernels.c.id])
+                .select_from(kernels)
+                .where(
+                    (kernels.c.session_id == session_id) &
+                    (kernels.c.status.in_(LIVE_STATUS))
+                )
+            )
+            result = await dbconn.execute(query)
+            rows = result.fetchall()
+            kernel_ids = [k["id"] for k in rows]
+        else:
+            kernel_ids = [session_id]
+        current_utilizations = await self.get_current_utilization(kernel_ids, occupied_slots)
+        if current_utilizations is None:
+            return True
+
+        # Update utilization time-series data.
+        not_enough_data = False
+        util_series_key = f"session.{session_id}.util_series"
+        raw_util_series = await self._redis.get(util_series_key, encoding=None)
+
+        try:
+            util_series = msgpack.unpackb(raw_util_series, use_list=True)
+        except TypeError:
+            util_series = {k: [] for k in self.resource_thresholds}
```

Comment:
To highlight that `self.resource_thresholds` is a dict, let's use `.keys()` here and in similar occurrences afterwards.

Reply:
using keys for self.resource_thresholds